homeassistant/components/input_timetable/__init__.py

```diff
@@ -0,0 +1,350 @@
+
+    @property
+    def state_attributes(self) -> dict:
+        """Return the state attributes."""
+        return {
+            ATTR_EDITABLE: self.editable,
+            ATTR_TIMETABLE: self._timetable_to_attribute(),
+        }
+
+    @property
+    def unique_id(self) -> str | None:
+        """Return unique id of the entity."""
+        return self._config[CONF_ID]
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state and state.attributes.get(ATTR_TIMETABLE):
+            self._timetable_from_attribute(state.attributes[ATTR_TIMETABLE])
+        self._update_state()
+
+    def _timetable_to_attribute(self) -> list:
+        return [
+            {ATTR_TIME: event.time.isoformat(), ATTR_STATE: event.state}
+            for event in self._timetable
+        ]
+
+    def _timetable_from_attribute(self, timetable: list) -> None:
+        self._timetable = [
+            StateEvent(
+                datetime.time.fromisoformat(event[ATTR_TIME]),
+                event[ATTR_STATE],
+            )
+            for event in timetable
+        ]
+        self._sort_timetable()
+
+    def _sort_timetable(self) -> None:
+        self._timetable.sort(key=lambda event: event.time)
+
+    async def async_set(self, time: datetime.time, state: str) -> None:
+        """Add a state change event to the timetable."""
+        for event in self._timetable:
+            if event.time == time:
+                event.state = state
+                break
+        else:
+            self._timetable.append(StateEvent(time, state))
+            self._sort_timetable()
+        self._update_state()
+
+    async def async_unset(self, time: datetime.time) -> None:
+        """Remove a state change event."""
+        for event in self._timetable:
+            if event.time == time:
+                self._timetable.remove(event)
+                break
+        else:
+            raise vol.Invalid(f"The time {time.isoformat()} doesn't exist")
+        self._update_state()
+
+    async def async_reset(self) -> None:
+        """Remove all state changes."""
+        self._timetable.clear()
+        self._update_state()
+
+    async def async_reconfig(self, timetable: list) -> None:
+        """Override the timetable with the new list."""
+        if len(timetable) > len({event[ATTR_TIME] for event in timetable}):
+            raise vol.Invalid("The same time is not allowed more than once")
+        self._timetable = [
+            StateEvent(event[ATTR_TIME], event[ATTR_STATE]) for event in timetable
+        ]
+        self._sort_timetable()
+        self._update_state()
+
+    async def async_update_config(self, config: dict) -> None:
+        """Handle when the config is updated."""
+        self._config = config
+        self._update_state()
+
+    def _schedule_update(self) -> None:
```

Comment:
this seems like very specific logic, can maybe add a bit of doc?

Reply:
Done